[IMP] Manufacturing: Resolving work order dependencies v19 (#17304)
odoo/documentation: content/applications/inventory_and_mrp/manufacturing/advanced_configuration/work_order_dependencies.rst

.. |BOM| replace:: :abbr:`BoM (Bill of Materials)`

When manufacturing certain products, specific operations may need to be completed before others can

> begin. To ensure operations are carried out in the correct order, the *Manufacturing* app features a

Reviewer: ```suggestion
begin. To ensure operations are carried out in the correct order, the **Manufacturing** app features a
```

*work order dependencies* setting. Enabling this setting allows for operations on a Bill of

> Materials (BoM) to be *blocked* by other operations that should occur first.

Reviewer: ```suggestion
*work order dependencies* setting. Enabling this setting allows for operations on a bill of
materials (BoM) to be blocked by other operations that should occur first.
```

:guilabel:`(settings)` menu of the *Operations* tab.

.. image:: work_order_dependencies/operation-dependencies.png
   :alt: The Operation Dependencies checkbox on the Miscellaneous tab of a BoM.

Open the *Operations* tab. Click the :icon:`oi-settings-adjust` :guilabel:`(settings)` icon, then
enable the :guilabel:`Blocked By` checkbox. The :guilabel:`Blocked By` field appears for each
operation in the tab.

.. image:: work_order_dependencies/operations-settings.png
   :alt: The settings for the Operations tab on a BoM.

In the line of the operation that should be blocked by another operation, click the
:guilabel:`Blocked By` field. The *Open: Operations* pop-up window opens. In the :guilabel:`Blocked
By` field, select the blocking operation that must be completed *before* the operation that is
blocked.

.. image:: work_order_dependencies/blocked-by.png
   :alt: The Blocked By drop-down field for an operation on a BoM.

Finally, save the |BOM| by clicking :icon:`fa-cloud-upload` :guilabel:`(Save manually)`.

.. seealso::
   :doc:`../basic_setup/bill_configuration`

Plan work orders using dependencies
===================================

> After work order dependencies are configured for a |BOM|, the *Manufacturing* app can plan when work

Reviewer: ```suggestion
After work order dependencies are configured for a |BOM|, the **Manufacturing** app can plan when work
```

orders are scheduled based on those dependencies. To plan the work orders for a manufacturing order,
navigate to :menuselection:`Manufacturing app --> Operations --> Manufacturing Orders`.

Next, select a manufacturing order for a product with work order dependencies set on its |BOM|, or
create a new manufacturing order by clicking :guilabel:`New`. If a new manufacturing order is
created, select a |BOM| configured with work order dependencies from the :guilabel:`Bill of
Material` drop-down field, then click :guilabel:`Confirm`.

> After confirming the manufacturing order, open the *Work Orders* tab to view the work orders

Reviewer: IIRC we decided that boldface can be used instead of italics if the title element is being interacted with, so i think that's an option if you want to emphasize the user action

- :menuselection:`Manufacturing app --> Planning --> Planning by Workcenter`

Both open a *Work Orders Planning* page. This page shows a timeline of all the work orders scheduled
for each operation.

:guilabel:`Planning by Production` is best used to view work orders grouped by manufacturing order.
This view is ideal for providing estimated delivery times or determining if an order will be
completed on schedule.

:guilabel:`Planning by Workcenter` is best used to view work orders grouped by the work center at
which they take place. This view can help identify overloaded machines or bottlenecks in production.

In both views, if one work order is blocked by the completion of another, the work order that is
blocked is shown as scheduled to start after the work order that blocks it. In addition, a grey
arrow connects the two work orders, leading from the blocking operation to the blocked operation.

.. image:: work_order_dependencies/planning-arrow.png
   :alt: The arrow connecting a blocked work order to the work order blocking it.

Rescheduling work orders
~~~~~~~~~~~~~~~~~~~~~~~~

By default, linked work orders on the *Work Orders Planning* page can be rescheduled automatically,

> ensuring that work orders cannot be moved out of the order they are set in.\

Reviewer 2: ```suggestion
ensuring that work orders cannot be moved out of the order they are set in.
```
typo?

> Planning by workcenter

Reviewer: ```suggestion
View work order scheduling by workcenter
```

Reviewer: 1. it seems like this subsection doesn't actually cover any workflow actions, just viewing the schedule? if that is the case i think the verb could be "view", to reflect our standard article structure of _intro > config > workflow > view/reporting/troubleshooting _, but specific wording up to you!
2. if this subsection is viewing only and not planning, i think it could be shifted up to the heading level

----------------------

To see a visual representation of how work orders are planned, navigate to the *Work Orders
Planning* page by going to :menuselection:`Manufacturing app --> Planning --> Planning by
Workcenter`. This page shows a timeline of all the work orders scheduled for each operation.

If one work order is blocked by the completion of another, the work order that is blocked is shown
as scheduled to start after the work order that blocks it. In addition, an arrow connects the two
work orders, leading from the blocking operation to the blocked operation.

.. image:: work_order_dependencies/planning-arrow.png
   :alt: The arrow connecting a blocked work order to the work order blocking it.

By default, linked work orders on the *Work Orders Planning* page can be rescheduled automatically,
ensuring that work orders cannot be moved out of the order they are set in.

> Resolving inconsistencies

Reviewer: can inconsistencies only be resolved by work center, via the _work orders planning_ page? wondering if it could be a main level heading as the "troubleshooting" section of the article

alt. based on my [next comment (line 116)](https://github.com/odoo/documentation/pull/17304/changes#r3058741432): if you do separate into _automatic vs. manual_ workflows, this could be a subsection under _manual_

~~~~~~~~~~~~~~~~~~~~~~~~~

> When the *Work Orders Planning* page is set to *manually reschedule* work orders, it is possible to

Reviewer: 1. how do you set it to _manually reschedule_?
2. if there are 2 discrete workflows based on auto-schedule vs. manual, consider reflecting that in headings

introduce inconsistencies in the manufacturing order.

The arrows on the *Work Orders Planning* page can also point out inconsistencies in the
manufacturing order and how to correct them.

.. example::

>    A workshop is manufacturing a table in three work orders: Assemble, apply finish, and dry.

Reviewer: maybe consider a numbered list?

   If the third work order (drying) is moved before the second work order, a red line appears to

>    link the two. The line is red to indicate that the moved work order violates the defined work

Reviewer: ```suggestion
   If the third work order (drying) is moved before the second work order, the third work order and the arrow pointing to it turn red to indicate that the planned schedule violates the defined work
```

Reviewer: IMO the bigger visual indicator is the WO in red!

changed "moved work order" bc "defined work order dependencies" is used right after, but wording up to you

>    order dependencies.

Reviewer: ```suggestion
   order dependency.
```

Reviewer: it's a singular dependency between the two WO's right? is there a way to set up recurring dependencies or multi-step dependencies?

   .. image:: work_order_dependencies/out-of-order.png
      :alt: A red line appears between two work orders.

When hovering over the connecting line, an **X** (cancel) button appears. When clicking this button,
Odoo assumes the dependencies no longer apply due to an exception, and the link between the steps is

> broken. Click this button to break that link.

Reviewer: ```suggestion
To remove the dependency, hover over the connecting line. Click the **X** (cancel) button that appears, and the arrow between the two work orders disappears.
```

To otherwise resolve the inconsistency, select the red work order, then click the :guilabel:`Edit`
button. The *Open* window opens. In this window, the :guilabel:`Start Date` has a red
:icon:`fa-exclamation-triangle` :guilabel:`(exclamation triangle)` icon next to it. Click this icon,

> and in the *Scheduling Information* pop-up, click the :guilabel:`Replan` button.

Reviewer: ```suggestion
To resolve the scheduling conflict and keep the dependency, select the red work order. Click the :guilabel:`Edit`
button, and the *Open* window appears. Click the red :icon:`fa-exclamation-triangle` :guilabel:`(exclamation triangle)` icon next to the the :guilabel:`Start Date`,
and the *Scheduling Information* pop-up window appears.
```

.. image:: work_order_dependencies/replan-work-order.png
   :alt: Click the Replan button to reschedule the inconsistent work order.

> Click :guilabel:`Save` to save the changes to the work order. The work order is rescheduled in a way

Reviewer: ```suggestion
Click the :guilabel:`Replan` button, then click :guilabel:`Save` to save the changes to the work order. The work order is rescheduled after the previous work order in a way
```

> that maintains the work order dependencies.

Reviewer: ```suggestion
that maintains the work order dependency.
```
